ds_test.go

```diff
@@ -0,0 +1,567 @@
+	orderByDesc := []dsq.Order{orbkd}
+	rs, err = d.Query(dsq.Query{Prefix: "/a/", Orders: orderByDesc})
+	if err != nil {
+		t.Fatal(err)
+	}
+	testKeyOrder(t, rs, []string{
+		"/a/d",
+		"/a/c",
+		"/a/b/d",
+		"/a/b/c",
+		"/a/b",
+	})
+
+	// test filters
+	equalFilter := dsq.FilterKeyCompare{Op: dsq.Equal, Key: "/a/b"}
+	equalFilters := []dsq.Filter{equalFilter}
+	rs, err = d.Query(dsq.Query{Prefix: "/a/", Filters: equalFilters})
+	if err != nil {
+		t.Fatal(err)
+	}
+	testKeyFilter(t, rs, []string{"/a/b"})
+
+	greaterThanFilter := dsq.FilterKeyCompare{Op: dsq.GreaterThan, Key: "/a/b"}
+	greaterThanFilters := []dsq.Filter{greaterThanFilter}
+	rs, err = d.Query(dsq.Query{Prefix: "/a/", Filters: greaterThanFilters})
+	if err != nil {
+		t.Fatal(err)
+	}
+	testKeyFilter(t, rs, []string{
+		"/a/b/c",
+		"/a/b/d",
+		"/a/c",
+		"/a/d",
+	})
+
+	lessThanFilter := dsq.FilterKeyCompare{Op: dsq.LessThanOrEqual, Key: "/a/b/c"}
+	lessThanFilters := []dsq.Filter{lessThanFilter}
+	rs, err = d.Query(dsq.Query{Prefix: "/a/", Filters: lessThanFilters})
+	if err != nil {
+		t.Fatal(err)
+	}
+	testKeyFilter(t, rs, []string{
+		"/a/b",
+		"/a/b/c",
+	})
+}
+
+func TestHas(t *testing.T) {
+	d, done := newDS(t)
+	defer done()
+	addTestCases(t, d, testcases)
+
+	has, err := d.Has(ds.NewKey("/a/b/c"))
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !has {
+		t.Error("Key should be found")
+	}
+
+	has, err = d.Has(ds.NewKey("/a/b/c/d"))
+	if err != nil {
+		t.Error(err)
+	}
+
+	if has {
+		t.Error("Key should not be found")
+	}
+}
+
+func TestNotExistGet(t *testing.T) {
+	d, done := newDS(t)
+	defer done()
+	addTestCases(t, d, testcases)
+
+	has, err := d.Has(ds.NewKey("/a/b/c/d"))
+	if err != nil {
+		t.Error(err)
+	}
+
+	if has {
+		t.Error("Key should not be found")
+	}
+
+	val, err := d.Get(ds.NewKey("/a/b/c/d"))
+	if val != nil {
+		t.Error("Key should not be found")
+	}
+
+	if err != ds.ErrNotFound {
+		t.Error("Error was not set to ds.ErrNotFound")
+		if err != nil {
+			t.Error(err)
+		}
+	}
+}
+
+func TestDelete(t *testing.T) {
+	d, done := newDS(t)
+	defer done()
+	addTestCases(t, d, testcases)
+
+	has, err := d.Has(ds.NewKey("/a/b/c"))
+	if err != nil {
+		t.Error(err)
+	}
+	if !has {
+		t.Error("Key should be found")
+	}
+
+	err = d.Delete(ds.NewKey("/a/b/c"))
+	if err != nil {
+		t.Error(err)
+	}
+
+	has, err = d.Has(ds.NewKey("/a/b/c"))
+	if err != nil {
+		t.Error(err)
+	}
+	if has {
+		t.Error("Key should not be found")
+	}
+}
+
+func TestGetEmpty(t *testing.T) {
+	d, done := newDS(t)
+	defer done()
+
+	err := d.Put(ds.NewKey("/a"), []byte{})
+	if err != nil {
+		t.Error(err)
+	}
+
+	v, err := d.Get(ds.NewKey("/a"))
+	if err != nil {
+		t.Error(err)
+	}
+
+	if len(v.([]byte)) != 0 {
+		t.Error("expected 0 len []byte form get")
+	}
+}
+
+func expectMatches(t *testing.T, expect []string, actualR dsq.Results) {
+	actual, err := actualR.Rest()
+	if err != nil {
+		t.Error(err)
+	}
+
+	if len(actual) != len(expect) {
+		fmt.Println("actual: ", actual)
```

Comment:
should probably remove `fmt.Println` now and add more informative message for error